### PR TITLE
[mlir][tensor] Update `GeneralizeOuterUnitDimsPackOpPattern`

### PR DESCRIPTION
**Pull request by banach-space**
Avoid generating spurious tensor.extract_slice, follow-on for #114315.

This is best to demonstrate with an example. Here's input for
`GeneralizeOuterUnitDimsPackOpPattern`:
```mlir
%pack = tensor.pack %input
  padding_value(%pad : f32)
  inner_dims_pos = [1, 0]
  inner_tiles = [2, %tile_dim_1]
  into %output : tensor<5x1xf32> -> tensor<1x1x2x?xf32>
```

Output _before_:
```mlir
%padded = tensor.pad %arg0 low[0, 0] high[%0, 1] {
^bb0(%arg4: index, %arg5: index):
  tensor.yield %arg2 : f32
} : tensor<5x1xf32> to tensor<?x2xf32>
// NOTE: skipped in the output _after_
%extracted_slice = tensor.extract_slice
  %padded[0, 0] [%arg3, 2] [1, 1] :
  tensor<?x2xf32> to tensor<?x2xf32>
%empty = tensor.empty(%arg3) : tensor<2x?xf32>
%transposed = linalg.transpose
  ins(%extracted_slice : tensor<?x2xf32>)
  outs(%empty : tensor<2x?xf32>)
  permutation = [1, 0]
%inserted_slice = tensor.insert_slice %transposed=
  into %arg1[0, 0, 0, 0] [1, 1, 2, %arg3] [1, 1, 1, 1] :
  tensor<2x?xf32> into tensor<1x1x2x?xf32>
```

Output _after_:
```mlir
%padded = tensor.pad %arg0 low[0, 0] high[%0, 1] {
^bb0(%arg4: index, %arg5: index):
  tensor.yield %arg2 : f32
} : tensor<5x1xf32> to tensor<?x2xf32>
%empty = tensor.empty(%arg3) : tensor<2x?xf32>
%transposed = linalg.transpose
  ins(%padded : tensor<?x2xf32>)
  outs(%empty : tensor<2x?xf32>) permutation = [1, 0]
%inserted_slice = tensor.insert_slice %transposed
  into %arg1[0, 0, 0, 0] [1, 1, 2, %arg3] [1, 1, 1, 1] :
  tensor<2x?xf32> into tensor<1x1x2x?xf32>
```

This PR also adds a check to verify that only the last N trailing dimensions
are tiled (for some value of N). Based on the PR discussion, this
restriction seems reasonable—especially as there are no in-tree tests
requiring otherwise. For now, it also simplifies the computation of
permutations for linalg.transpose. This restriction can be relaxed in
the future if needed.

### CODE REVIEW

**Review by Max191 — APPROVED** [review]
This looks good to me, but I'm not sure that this assumption about tensor.pack op is actually true in practice:
> This PR also adds a check to verify that only the last N (for some value
of N) trailing dims that are being tiled. From what I can tell, that's
always the case in practice.

I have seen pack ops show up in practice that do not follow this restriction, so it may be worth supporting this case, but I'm not sure we rely much on this pattern anymore anyway. I'll approve, but pinging @hanhanW, who might have a better idea of whether or not support for this case would be desired as a follow up.

**Comment by hanhanW** [comment]
> > This PR also adds a check to verify that only the last N (for some value
> > of N) trailing dims that are being tiled. From what I can tell, that's
> > always the case in practice.
> 
> I have seen pack ops show up in practice that do not follow this restriction, so it may be worth supporting this case, but I'm not sure we rely much on this pattern anymore anyway. I'll approve, but pinging @hanhanW, who might have a better idea of whether or not support for this case would be desired as a follow up.

It is not a restriction in our use cases because we can data-tile any ops that implements ContractionOpInterface. E.g., a contraction generic op which has batch dimension being the innermost dimension. In this case, we don't data-tile batch dimension.

These patterns were built for the first take of pack/unpack vectorization, and they are no longer used on IREE CPU x86 codegen path. Because today we have direct vectorization and masking supports.

**Comment by banach-space** [comment]
> It is not a restriction in our use cases because we can data-tile any ops that implements ContractionOpInterface. E.g., a contraction generic op which has batch dimension being the innermost dimension. In this case, we don't data-tile batch dimension.
> 
> These patterns were built for the first take of pack/unpack vectorization, and they are no longer used on IREE CPU x86 codegen path. Because today we have direct vectorization and masking supports.

Thanks for the context and for the explanation! Any preference how to proceed here?

From my perspective, without tests in-tree, it tends to be tricky to produce good ref examples. So, if I can simplify things without breaking any tests, I'd go with that.

**Review by hanhanW — APPROVED** [review]
LG, just a couple nits.

File: mlir/test/Dialect/Linalg/generalize-tensor-pack-tile.mlir [hunk 1]
```diff
@@ -9,19 +9,19 @@ func.func @KCRS_to_KCRSsr(%arg0: tensor<1x1x128x64xf32>, %arg1: tensor<1x1x4x8x8
...
 // CHECK-SAME:          [0, 0, %[[IN_R]], %[[IN_S]]] [1, 1, 32, 8] [1, 1, 1, 1]
-// CHECK:             %[[TILE:.+]] = tensor.extract_slice %[[SRC_SLICE]]
-// CHECK-SAME:          [0, 0, 0, 0] [1, 1, 32, 8] [1, 1, 1, 1] : tensor<1x1x32x8xf32> to tensor<32x8xf32>
-// CHECK:             %[[EMPTY:.+]] = tensor.empty() : tensor<8x32xf32>
-// CHECK:             %[[TRANSP:.+]] =  linalg.transpose
-// CHECK-SAME:          ins(%[[TILE]]
-// CHECK-SAME:          outs(%[[EMPTY]]
-// CHECK-SAME:          permutation = [1, 0]
+// CHECK:             %[[TILE:.*]] = tensor.extract_slice %[[ITER_SLICE]]
+// CHECK-SAME:          [0, 0, %[[R]], %[[S]], 0, 0] [1, 1, 1, 1, 8, 32] [1, 1, 1, 1, 1, 1] : tensor<1x1x4x8x8x32xf32> to tensor<1x1x1x1x8x32xf32>
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Linalg/generalize-tensor-pack-tile.mlir:19]: Yeah, it's a tile for `tensor.insert_slice` below. We are not capturing in the expected output _before_, so I shouldn't be capturing it _after_. Let me remove it, thanks for catching this!

**Review by hanhanW — APPROVED** [review]
LGTM, thanks

**Pull request merged**